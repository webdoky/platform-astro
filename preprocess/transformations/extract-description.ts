import type { Root, Yaml } from 'mdast';
import { toString } from 'mdast-util-to-string';
import { EXIT, visit } from 'unist-util-visit';

import setFrontmatterField from './utils/set-frontmatter-field.js';
import stripMacros from './utils/strip-macros.js';

export default function extractDescription(tree: Root) {
  const frontmatterNode = tree.children.find(
    (node) => node.type === 'yaml',
  ) as Yaml;
  if (!frontmatterNode) {
    console.warn('No frontmatter found');
    return;
  }

  visit(tree, 'paragraph', (node) => {
    let text = toString(node).trim();
    // strip HTML from the text
    text = text.replaceAll(/<[^>]*>/g, '');
    // strip macros from the text
    text = stripMacros(text);
    if (text && !text.startsWith('<') && !text.startsWith('{{')) {
      setFrontmatterField(frontmatterNode, 'description', text);
      return [EXIT];
    }
    return;
  });
}

export function extractDescriptionPlugin() {
  return extractDescription;
}
