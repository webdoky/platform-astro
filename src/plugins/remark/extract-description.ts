import type { Root } from 'mdast';
import { toString } from 'mdast-util-to-string';
import { EXIT, visit } from 'unist-util-visit';

import stripMacros from '../utils/strip-macros.js';

import createRemarkPlugin from './create-plugin.js';
import type { AstroFile } from './validate-astro-file.ts';

export default function extractDescription(tree: Root, file: AstroFile) {
  visit(tree, 'paragraph', (node) => {
    let text = toString(node).trim();
    // strip HTML from the text
    text = text.replaceAll(/<[^>]*>/g, '');
    // strip macros from the text
    text = stripMacros(text);
    if (text && !text.startsWith('<') && !text.startsWith('{{')) {
      file.data.astro.frontmatter.description = text;
      return [EXIT];
    }
    return;
  });
}

export const extractDescriptionPlugin = createRemarkPlugin(extractDescription);
