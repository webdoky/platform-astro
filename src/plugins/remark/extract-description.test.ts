import type { Root } from 'mdast';

import { extractDescriptionPlugin } from './extract-description.js';
import type { AstroFile } from './validate-astro-file.ts';

describe('extractDescription', () => {
  test('sets the description and sets it in the frontmatter', () => {
    const tree: Root = {
      type: 'root',
      children: [
        {
          type: 'paragraph',
          children: [
            {
              type: 'text',
              value: 'description',
            },
          ],
        },
      ],
    };
    const file = {
      data: {
        astro: {
          frontmatter: {
            slug: 'slug',
            title: 'title',
          },
        },
      },
    };

    extractDescriptionPlugin()(tree, file as unknown as AstroFile);
    expect(file).toMatchSnapshot();
  });
});
