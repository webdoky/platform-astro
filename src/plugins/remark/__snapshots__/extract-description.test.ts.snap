// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extractDescription > sets the description and sets it in the frontmatter 1`] = `
{
  "data": {
    "astro": {
      "frontmatter": {
        "description": "description",
        "slug": "slug",
        "title": "title",
      },
    },
  },
}
`;
